frontend/components/AgentWallet/FundAgent/useFundAgent.tsx

```diff
@@ -0,0 +1,123 @@
+import { sum } from 'lodash';
+import { useCallback, useMemo, useState } from 'react';
+import { useUnmount } from 'usehooks-ts';
+
+import { CHAIN_CONFIG } from '@/config/chains';
+import { TOKEN_CONFIG, TokenConfig } from '@/config/tokens';
+import { TokenSymbol, TokenSymbolMap } from '@/constants';
+import {
+  useBalanceContext,
+  useMasterBalances,
+  useRewardContext,
+  useService,
+  useServices,
+  useUsdAmounts,
+} from '@/hooks';
+import { asEvmChainDetails } from '@/utils';
+
+import { AvailableAsset } from '../types';
+
+export const useFundAgent = () => {
+  const { isLoading: isBalanceLoading } = useBalanceContext();
+  const { accruedServiceStakingRewards } = useRewardContext();
+  const {
+    masterEoaNativeBalance,
+    masterSafeNativeBalance,
+    masterSafeOlasBalance,
+    masterSafeErc20Balances,
+  } = useMasterBalances();
+  const {
+    isLoading: isServicesLoading,
+    selectedAgentConfig,
+    selectedService,
+  } = useServices();
+  const { isLoaded } = useService(selectedService?.service_config_id);
+
+  const { evmHomeChainId, middlewareHomeChainId } = selectedAgentConfig;
+
+  const [amountsToWithdraw, setAmountsToWithdraw] = useState<
+    Partial<Record<TokenSymbol, number>>
+  >({});
+
+  const usdRequirements = useMemo(() => {
+    if (!evmHomeChainId) return [];
+    return Object.entries(TOKEN_CONFIG[evmHomeChainId!]).map(
+      ([untypedSymbol]) => {
+        const symbol = untypedSymbol as TokenSymbol;
+        return { symbol, amount: 0 };
+      },
+    );
+  }, [evmHomeChainId]);
+
+  const chainName = evmHomeChainId
+    ? (CHAIN_CONFIG[evmHomeChainId].name as string)
+    : '';
+  const { breakdown: usdBreakdown } = useUsdAmounts(chainName, usdRequirements);
+
+  // OLAS token, Native Token, other ERC20 tokens
+  const availableAssets: AvailableAsset[] = useMemo(() => {
```

Comment:
is this copied from PearlWalletProvider? what's the purpose of it then and why it can't be used here? maybe we can share some logic if you see clear separation?